crates/cubecl-opt/src/analyses/base.rs

```diff
@@ -0,0 +1,62 @@
+use std::{any::Any, cell::RefCell, rc::Rc};
+
+use type_map::TypeMap;
+
+use crate::Optimizer;
+
+use super::{
+    dominance::{Dominators, PostDominators},
+    liveness::Liveness,
+    post_order::PostOrder,
+};
+
+/// An analysis used by optimization passes. Unlike optimization passes, analyses can have state
+/// and persist until they're invalidated.
+pub trait Analysis {
+    /// Perform the analysis for the current optimizer state and return the persistent analysis state
+    fn init(opt: &mut Optimizer) -> Self;
+}
+
+#[derive(Default, Clone, Debug)]
+pub struct Analyses {
+    cache: Rc<RefCell<TypeMap>>,
+}
+
+impl Analyses {
+    pub fn get<A: Analysis + Any>(&self, opt: &mut Optimizer) -> Rc<A> {
+        let analysis = self.cache.borrow().get::<Rc<A>>().cloned();
+        if let Some(analysis) = analysis {
+            analysis
+        } else {
+            let analysis = Rc::new(A::init(opt));
+            self.cache.borrow_mut().insert(analysis.clone());
+            analysis
+        }
+    }
+
+    pub fn try_get<A: Any>(&self) -> Option<Rc<A>> {
+        self.cache.borrow().get().cloned()
+    }
+
+    pub fn invalidate<A: Analysis + Any>(&self) {
+        self.cache.borrow_mut().remove::<Rc<A>>();
+    }
+}
+
+impl Optimizer {
+    pub fn analysis<A: Analysis + Any>(&mut self) -> Rc<A> {
+        let analyses = self.analyses.clone();
+        analyses.get(self)
+    }
+
+    pub fn invalidate_analysis<A: Analysis + Any>(&self) {
+        self.analyses.invalidate::<A>();
+    }
+
+    pub fn invalidate_structure(&self) {
+        self.invalidate_analysis::<PostOrder>();
+        self.invalidate_analysis::<Dominators>();
+        self.invalidate_analysis::<PostDominators>();
+        self.invalidate_analysis::<Liveness>();
+    }
```

Comment:
Could have some basic docs. Like is there a reason why we fetch the anlysis from the optimizer using it's type?

Reply:
I mean I like it, just some clarifications would be great!

Reply:
I added some basic docs, but I figured the type was obvious - different analyses provide different data and methods, and they're not defined by a single kind of access. So they need to be stored by their type ID and cast to the specific type. Analyses are always singletons so the type itself can serve as the key.